// ngx-material-time-select/esm5/lib/moment-time-adapter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Inject, Injectable, Optional } from '@angular/core';
import { DateAdapter, MAT_DATE_LOCALE } from '@angular/material';
import { TimeAdapter } from './time-adapter.service';
import * as _moment from 'moment';
/** @type {?} */
var moment = _moment;
/**
 * Adapts Moment.js Dates for use with Angular Material.
 */
var MomentTimeAdapter = /** @class */ (function (_super) {
    tslib_1.__extends(MomentTimeAdapter, _super);
    function MomentTimeAdapter(locale, _dateAdapter) {
        var _this = _super.call(this) || this;
        _this._dateAdapter = _dateAdapter;
        _this.setLocale(locale || moment.locale());
        return _this;
    }
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.getYear = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.clone(date).year();
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.getMonth = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.clone(date).month();
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.getDate = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.clone(date).date();
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.getDayOfWeek = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.clone(date).day();
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.getHour = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.clone(date).hour();
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.getMinute = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.clone(date).minute();
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.getSecond = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.clone(date).second();
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.clone = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return date.clone().locale(this.locale);
    };
    /**
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.createDate = /**
     * @param {?} year
     * @param {?} month
     * @param {?} date
     * @return {?}
     */
    function (year, month, date) {
        return this.clone(this._dateAdapter.createDate(year, month, date));
    };
    /**
     * @param {?=} hour
     * @param {?=} minute
     * @param {?=} second
     * @return {?}
     */
    MomentTimeAdapter.prototype.createTime = /**
     * @param {?=} hour
     * @param {?=} minute
     * @param {?=} second
     * @return {?}
     */
    function (hour, minute, second) {
        if (hour === void 0) { hour = 0; }
        if (minute === void 0) { minute = 0; }
        if (second === void 0) { second = 0; }
        if (hour < 0 || hour > 23) {
            throw new Error("Invalid hour " + hour + ". Hour has to be between 0 and 23.");
        }
        if (minute < 0 || minute > 59) {
            throw new Error("Invalid minute " + minute + ". Minute has to be between 0 and 59.");
        }
        if (second < 0 || second > 59) {
            throw new Error("Invalid second " + second + ". Second has to be between 0 and 59.");
        }
        return moment({ hour: hour, minute: minute, second: second }).locale(this.locale);
    };
    /**
     * @return {?}
     */
    MomentTimeAdapter.prototype.today = /**
     * @return {?}
     */
    function () {
        return this.now();
    };
    /**
     * @return {?}
     */
    MomentTimeAdapter.prototype.now = /**
     * @return {?}
     */
    function () {
        return moment().locale(this.locale);
    };
    /**
     * @param {?} value
     * @param {?} parseFormat
     * @return {?}
     */
    MomentTimeAdapter.prototype.parse = /**
     * @param {?} value
     * @param {?} parseFormat
     * @return {?}
     */
    function (value, parseFormat) {
        if (value && typeof value === 'string') {
            return moment(value, parseFormat, this.locale);
        }
        return value ? moment(value).locale(this.locale) : null;
    };
    /**
     * @param {?} date
     * @param {?} displayFormat
     * @return {?}
     */
    MomentTimeAdapter.prototype.format = /**
     * @param {?} date
     * @param {?} displayFormat
     * @return {?}
     */
    function (date, displayFormat) {
        date = this.clone(date);
        if (!this.isValid(date)) {
            throw new Error('MomentTimeAdapter: Cannot format invalid date.');
        }
        return date.format(displayFormat);
    };
    /**
     * @param {?} date
     * @param {?} years
     * @return {?}
     */
    MomentTimeAdapter.prototype.addCalendarYears = /**
     * @param {?} date
     * @param {?} years
     * @return {?}
     */
    function (date, years) {
        return this.clone(date).add({ years: years });
    };
    /**
     * @param {?} date
     * @param {?} months
     * @return {?}
     */
    MomentTimeAdapter.prototype.addCalendarMonths = /**
     * @param {?} date
     * @param {?} months
     * @return {?}
     */
    function (date, months) {
        return this.clone(date).add({ months: months });
    };
    /**
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    MomentTimeAdapter.prototype.addCalendarDays = /**
     * @param {?} date
     * @param {?} days
     * @return {?}
     */
    function (date, days) {
        return this.clone(date).add({ days: days });
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.toIso8601 = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.clone(date).format();
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.toMoment = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.clone(date);
    };
    /**
     * @param {?} value
     * @return {?}
     */
    MomentTimeAdapter.prototype.deserialize = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        /** @type {?} */
        var date;
        if (value instanceof Date) {
            date = moment(value);
        }
        if (typeof value === 'string') {
            if (!value) {
                return null;
            }
            date = moment(value, moment.ISO_8601);
        }
        if (date && this.isValid(date)) {
            return date;
        }
        return _super.prototype.deserialize.call(this, value);
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    MomentTimeAdapter.prototype.isDateInstance = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        return moment.isMoment(obj);
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.isValid = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return this.clone(date).isValid();
    };
    /**
     * @return {?}
     */
    MomentTimeAdapter.prototype.invalid = /**
     * @return {?}
     */
    function () {
        return moment.invalid();
    };
    /**
     * @return {?}
     */
    MomentTimeAdapter.prototype.getDateNames = /**
     * @return {?}
     */
    function () {
        return [];
    };
    /**
     * @param {?} style
     * @return {?}
     */
    MomentTimeAdapter.prototype.getDayOfWeekNames = /**
     * @param {?} style
     * @return {?}
     */
    function (style) {
        return [];
    };
    /**
     * @return {?}
     */
    MomentTimeAdapter.prototype.getFirstDayOfWeek = /**
     * @return {?}
     */
    function () {
        return 0;
    };
    /**
     * @param {?} style
     * @return {?}
     */
    MomentTimeAdapter.prototype.getMonthNames = /**
     * @param {?} style
     * @return {?}
     */
    function (style) {
        return [];
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.getNumDaysInMonth = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return 0;
    };
    /**
     * @param {?} date
     * @return {?}
     */
    MomentTimeAdapter.prototype.getYearName = /**
     * @param {?} date
     * @return {?}
     */
    function (date) {
        return '';
    };
    MomentTimeAdapter.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    MomentTimeAdapter.ctorParameters = function () { return [
        { type: String, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_LOCALE,] }] },
        { type: DateAdapter, decorators: [{ type: Optional }] }
    ]; };
    return MomentTimeAdapter;
}(TimeAdapter));
export { MomentTimeAdapter };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MomentTimeAdapter.prototype._dateAdapter;
}
