// ngx-material-time-select/esm5/lib/time-adapter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DateAdapter } from '@angular/material';
/**
 * Adapts date object to be usable as a date time by cdk-based components that work with dates.
 * @abstract
 * @template D
 */
var /**
 * Adapts date object to be usable as a date time by cdk-based components that work with dates.
 * @abstract
 * @template D
 */
TimeAdapter = /** @class */ (function (_super) {
    tslib_1.__extends(TimeAdapter, _super);
    function TimeAdapter() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Compares two date times.
     * @param first The first date to compare.
     * @param second The second date to compare.
     * @returns 0 if the date times are equal, a number less than 0 if the first date time is earlier,
     *     a number greater than 0 if the first date time is later.
     */
    /**
     * Compares two date times.
     * @param {?} first The first date to compare.
     * @param {?} second The second date to compare.
     * @return {?} 0 if the date times are equal, a number less than 0 if the first date time is earlier,
     *     a number greater than 0 if the first date time is later.
     */
    TimeAdapter.prototype.compareTime = /**
     * Compares two date times.
     * @param {?} first The first date to compare.
     * @param {?} second The second date to compare.
     * @return {?} 0 if the date times are equal, a number less than 0 if the first date time is earlier,
     *     a number greater than 0 if the first date time is later.
     */
    function (first, second) {
        return this.compareDate(first, second) ||
            this.getHour(first) - this.getHour(second) ||
            this.getMinute(first) - this.getMinute(second) ||
            this.getSecond(first) - this.getSecond(second);
    };
    /**
     * Checks if two date times are equal.
     * @param first The first date to check.
     * @param second The second date to check.
     * @returns Whether the two date times are equal.
     *     Null dates are considered equal to other null dates.
     */
    /**
     * Checks if two date times are equal.
     * @param {?} first The first date to check.
     * @param {?} second The second date to check.
     * @return {?} Whether the two date times are equal.
     *     Null dates are considered equal to other null dates.
     */
    TimeAdapter.prototype.sameTime = /**
     * Checks if two date times are equal.
     * @param {?} first The first date to check.
     * @param {?} second The second date to check.
     * @return {?} Whether the two date times are equal.
     *     Null dates are considered equal to other null dates.
     */
    function (first, second) {
        if (first && second) {
            /** @type {?} */
            var firstValid = this.isValid(first);
            /** @type {?} */
            var secondValid = this.isValid(second);
            if (firstValid && secondValid) {
                return !this.compareTime(first, second);
            }
            return firstValid === secondValid;
        }
        return first === second;
    };
    /**
     * Clamp the given date between min and max date times.
     * @param date The date to clamp.
     * @param min The minimum value to allow. If null or omitted no min is enforced.
     * @param max The maximum value to allow. If null or omitted no max is enforced.
     * @returns `min` if `date` is less than `min`, `max` if date is greater than `max`,
     *     otherwise `date`.
     */
    /**
     * Clamp the given date between min and max date times.
     * @param {?} date The date to clamp.
     * @param {?=} min The minimum value to allow. If null or omitted no min is enforced.
     * @param {?=} max The maximum value to allow. If null or omitted no max is enforced.
     * @return {?} `min` if `date` is less than `min`, `max` if date is greater than `max`,
     *     otherwise `date`.
     */
    TimeAdapter.prototype.clampTime = /**
     * Clamp the given date between min and max date times.
     * @param {?} date The date to clamp.
     * @param {?=} min The minimum value to allow. If null or omitted no min is enforced.
     * @param {?=} max The maximum value to allow. If null or omitted no max is enforced.
     * @return {?} `min` if `date` is less than `min`, `max` if date is greater than `max`,
     *     otherwise `date`.
     */
    function (date, min, max) {
        if (min && this.compareTime(date, min) < 0) {
            return min;
        }
        if (max && this.compareTime(date, max) > 0) {
            return max;
        }
        return date;
    };
    return TimeAdapter;
}(DateAdapter));
/**
 * Adapts date object to be usable as a date time by cdk-based components that work with dates.
 * @abstract
 * @template D
 */
export { TimeAdapter };
if (false) {
    /**
     * Gets the hour component of the given date.
     * @abstract
     * @param {?} date The date to extract the hour from.
     * @return {?} The hour component (0-indexed, 0 = midnight).
     */
    TimeAdapter.prototype.getHour = function (date) { };
    /**
     * Gets the minute component of the given date.
     * @abstract
     * @param {?} date The date to extract the minute from.
     * @return {?} The minute component (0-indexed, 0 = start of hour).
     */
    TimeAdapter.prototype.getMinute = function (date) { };
    /**
     * Gets the second component of the given date.
     * @abstract
     * @param {?} date The date to extract the second from.
     * @return {?} The second component (0-indexed, 0 = start of minute).
     */
    TimeAdapter.prototype.getSecond = function (date) { };
    /**
     * Create a date with the given hour, minute and second. Does not allow over/under-flow of the
     * hour, minute and second.
     * @abstract
     * @param {?=} hour The hour of the date. Must be an integer 0 - 23.
     * @param {?=} minute The minute of the date. Must be an integer 0 - 59.
     * @param {?=} second The second of the date. Must be an integer 0 - 59.
     * @return {?} The new date.
     */
    TimeAdapter.prototype.createTime = function (hour, minute, second) { };
    /**
     * Gets the current date.
     * @abstract
     * @return {?} The current date.
     */
    TimeAdapter.prototype.now = function () { };
    /**
     * Convert the given date to a Moment object.
     * @abstract
     * @param {?} date the date object to convert.
     * @return {?}
     */
    TimeAdapter.prototype.toMoment = function (date) { };
}
