// ngx-material-time-select/esm5/lib/time-select-toggle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Attribute, ChangeDetectionStrategy, ChangeDetectorRef, Component, ContentChild, Directive, HostBinding, HostListener, Input, ViewChild, ViewEncapsulation } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { MatButton } from '@angular/material';
import { merge, of, Subscription } from 'rxjs';
import { MatTimeSelectIntl } from './time-select-intl.service';
import { MatTimeSelectComponent } from './time-select.component';
/**
 * Can be used to override the icon of a `matTimeSelectToggle`.
 */
var MatTimeSelectToggleIconDirective = /** @class */ (function () {
    function MatTimeSelectToggleIconDirective() {
    }
    MatTimeSelectToggleIconDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[matTimeSelectToggleIcon]'
                },] }
    ];
    return MatTimeSelectToggleIconDirective;
}());
export { MatTimeSelectToggleIconDirective };
/**
 * @template D
 */
var MatTimeSelectToggleComponent = /** @class */ (function () {
    function MatTimeSelectToggleComponent(_intl, _changeDetectorRef, defaultTabIndex) {
        this._intl = _intl;
        this._changeDetectorRef = _changeDetectorRef;
        this._stateChanges = Subscription.EMPTY;
        /** @type {?} */
        var parsedTabIndex = parseInt(defaultTabIndex, 10);
        this.tabIndex = (parsedTabIndex || parsedTabIndex === 0) ? parsedTabIndex : null;
    }
    Object.defineProperty(MatTimeSelectToggleComponent.prototype, "disabled", {
        /** Whether the toggle button is disabled. */
        get: /**
         * Whether the toggle button is disabled.
         * @return {?}
         */
        function () {
            return this._disabled === undefined ? this.timeSelect.disabled : this._disabled;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimeSelectToggleComponent.prototype, "disableRipple", {
        /** Whether ripples on the toggle should be disabled. */
        get: /**
         * Whether ripples on the toggle should be disabled.
         * @return {?}
         */
        function () {
            return this._disableRipple;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disableRipple = coerceBooleanProperty(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimeSelectToggleComponent.prototype, "class", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return 'mat-time-select-toggle'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimeSelectToggleComponent.prototype, "tabindex", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return '-1'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimeSelectToggleComponent.prototype, "active", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.timeSelect && this.timeSelect.opened; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimeSelectToggleComponent.prototype, "accent", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.timeSelect && this.timeSelect.color === 'accent'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MatTimeSelectToggleComponent.prototype, "warn", {
        /** @docs-private */
        get: /**
         * \@docs-private
         * @return {?}
         */
        function () { return this.timeSelect && this.timeSelect.color === 'warn'; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    MatTimeSelectToggleComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this._watchStateChanges();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    MatTimeSelectToggleComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.timeSelect) {
            this._watchStateChanges();
        }
    };
    /**
     * @return {?}
     */
    MatTimeSelectToggleComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._stateChanges.unsubscribe();
    };
    /** Handle focus event. */
    /**
     * Handle focus event.
     * @return {?}
     */
    MatTimeSelectToggleComponent.prototype.onFocus = /**
     * Handle focus event.
     * @return {?}
     */
    function () {
        this.button.focus();
    };
    /**
     * Open the time select.
     * @param event The triggered event to open the time select.
     */
    /**
     * Open the time select.
     * @param {?} event The triggered event to open the time select.
     * @return {?}
     */
    MatTimeSelectToggleComponent.prototype.open = /**
     * Open the time select.
     * @param {?} event The triggered event to open the time select.
     * @return {?}
     */
    function (event) {
        if (this.timeSelect && !this.disabled) {
            this.timeSelect.open();
            event.stopPropagation();
        }
    };
    /**
     * @private
     * @return {?}
     */
    MatTimeSelectToggleComponent.prototype._watchStateChanges = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var timeSelectDisabled = this.timeSelect ? this.timeSelect.disabledChange : of();
        /** @type {?} */
        var inputDisabled = this.timeSelect && this.timeSelect._timeSelectInput ?
            this.timeSelect._timeSelectInput.disabledChange : of();
        /** @type {?} */
        var timeSelectToggled = this.timeSelect ?
            merge(this.timeSelect.openStream, this.timeSelect.closeStream) :
            of();
        this._stateChanges.unsubscribe();
        this._stateChanges = merge(this._intl.changes, timeSelectDisabled, inputDisabled, timeSelectToggled).subscribe(function () { return _this._changeDetectorRef.markForCheck(); });
    };
    MatTimeSelectToggleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'mat-time-select-toggle',
                    template: "<button\r\n  #button\r\n  mat-icon-button\r\n  type=\"button\"\r\n  aria-haspopup=\"true\"\r\n  [attr.aria-label]=\"_intl.openTimeSelectLabel\"\r\n  [attr.tabindex]=\"disabled ? -1 : tabIndex\"\r\n  [disabled]=\"disabled\"\r\n  [disableRipple]=\"disableRipple\"\r\n  (click)=\"open($event)\">\r\n\r\n  <svg\r\n    *ngIf=\"!customIcon\"\r\n    xmlns=\"http://www.w3.org/2000/svg\"\r\n    class=\"mat-time-select-toggle-default-icon\"\r\n    viewBox=\"0 0 24 24\"\r\n    width=\"24px\"\r\n    height=\"24px\"\r\n    fill=\"currentColor\"\r\n    focusable=\"false\">\r\n    <path d=\"M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z\"/>\r\n    <path d=\"M0 0h24v24H0z\" fill=\"none\"/>\r\n    <path d=\"M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z\"/>\r\n  </svg>\r\n\r\n  <ng-content select=\"[matTimeSelectToggleIcon]\"></ng-content>\r\n</button>\r\n",
                    exportAs: 'matTimeSelectToggle',
                    encapsulation: ViewEncapsulation.None,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: [".mat-form-field-appearance-legacy .mat-form-field-prefix .mat-time-select-toggle-default-icon,.mat-form-field-appearance-legacy .mat-form-field-suffix .mat-time-select-toggle-default-icon{width:1em}.mat-form-field:not(.mat-form-field-appearance-legacy) .mat-form-field-prefix .mat-time-select-toggle-default-icon,.mat-form-field:not(.mat-form-field-appearance-legacy) .mat-form-field-suffix .mat-time-select-toggle-default-icon{display:block;width:1.5em;height:1.5em}.mat-form-field:not(.mat-form-field-appearance-legacy) .mat-form-field-prefix .mat-icon-button .mat-time-select-toggle-default-icon,.mat-form-field:not(.mat-form-field-appearance-legacy) .mat-form-field-suffix .mat-icon-button .mat-time-select-toggle-default-icon{margin:auto}"]
                }] }
    ];
    /** @nocollapse */
    MatTimeSelectToggleComponent.ctorParameters = function () { return [
        { type: MatTimeSelectIntl },
        { type: ChangeDetectorRef },
        { type: String, decorators: [{ type: Attribute, args: ['tabindex',] }] }
    ]; };
    MatTimeSelectToggleComponent.propDecorators = {
        timeSelect: [{ type: Input, args: ['for',] }],
        tabIndex: [{ type: Input }],
        customIcon: [{ type: ContentChild, args: [MatTimeSelectToggleIconDirective,] }],
        button: [{ type: ViewChild, args: ['button',] }],
        disabled: [{ type: Input }],
        disableRipple: [{ type: Input }],
        class: [{ type: HostBinding, args: ['class',] }],
        tabindex: [{ type: HostBinding, args: ['attr.tabindex',] }],
        active: [{ type: HostBinding, args: ['class.mat-time-select-toggle-active',] }],
        accent: [{ type: HostBinding, args: ['class.mat-accent',] }],
        warn: [{ type: HostBinding, args: ['class.mat-warn',] }],
        onFocus: [{ type: HostListener, args: ['focus',] }]
    };
    return MatTimeSelectToggleComponent;
}());
export { MatTimeSelectToggleComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MatTimeSelectToggleComponent.prototype._stateChanges;
    /**
     * @type {?}
     * @private
     */
    MatTimeSelectToggleComponent.prototype._disabled;
    /**
     * @type {?}
     * @private
     */
    MatTimeSelectToggleComponent.prototype._disableRipple;
    /**
     * Time select instance that the button will toggle.
     * @type {?}
     */
    MatTimeSelectToggleComponent.prototype.timeSelect;
    /**
     * Tabindex for the toggle.
     * @type {?}
     */
    MatTimeSelectToggleComponent.prototype.tabIndex;
    /**
     * Custom icon set by the consumer.
     * @type {?}
     */
    MatTimeSelectToggleComponent.prototype.customIcon;
    /**
     * Underlying button element.
     * @type {?}
     */
    MatTimeSelectToggleComponent.prototype.button;
    /** @type {?} */
    MatTimeSelectToggleComponent.prototype._intl;
    /**
     * @type {?}
     * @private
     */
    MatTimeSelectToggleComponent.prototype._changeDetectorRef;
}
